Do not return `<none>` placeholders for images any more (#4673)
cri-o/cri-o: cri.md

# Container Runtime Interface special cases

> Target of this document is to outline corner cases and common pitfalls in

Reviewer: ```suggestion
The target of this document is to outline corner cases and common pitfalls in
```

conjunction with the Container Runtime Interface (CRI). This document outlines

> CRI-O's interpretation of certain aspects of the interface, which may be not

Reviewer: ```suggestion
CRI-O's interpretation of certain aspects of the interface, which may not be 
```

completely formalized.

The main documentation of the CRI can be found [in the corresponding protobuf
definition][0], whereas this document follows it on the `service`/`rpc` level.

## `ListImages`

> `ListImages` should lists existing images. Its response consists of an array of

Reviewer 2: ```suggestion
`ListImages` lists existing images. Its response consists of an array of
```

> `Image` types. Beside other information, an `Iamge` contains `repo_tags` and

Reviewer 2: ```suggestion
`Image` types. Besides other information, an `Image` contains `repo_tags` and
```

`repo_digests`, which are defined as:

```proto
// Other names by which this image is known.
repeated string repo_tags = 2;

// Digests by which this image is known.
repeated string repo_digests = 3;
```

> Both, tags and digests, will be used by:

Reviewer: ```suggestion
Both tags and digests will be used by:
```

> - The kubelet, which displays them in the node status as flat list, for example:

Reviewer: ```suggestion
- The kubelet, which displays them in the node status as a flat list, for example:
```

  ```
  > kubectl get node 127.0.0.1 -o json | jq .status.images
  ```

  ```json
  [
    {
      "names": [
        "k8s.gcr.io/pause@sha256:4a1c4b21597c1b4415bdbecb28a3296c6b5e23ca4f9feeb599860a1dac6a0108",
        "k8s.gcr.io/pause@sha256:927d98197ec1141a368550822d18fa1c60bdae27b78b0c004f705f548c07814f",
        "k8s.gcr.io/pause:3.2"
      ],
      "sizeBytes": 688049
    }
  ]
  ```

>   Right now the amount of images is limited by the kubelet flag

Reviewer 2: ```suggestion
  Right now, the amount of images is limited by the kubelet flag
```

  `--node-status-max-images` (). The scheduler uses this list to
  score nodes based on the information if a container image already exists.

- crictl, which is able to output the image list in a human readable way:
  ```
  > sudo crictl images --digests
  IMAGE                 TAG       DIGEST           IMAGE ID         SIZE
  k8s.gcr.io/pause      3.2       4a1c4b21597c1    80d28bedfe5de    688kB
  ```

> CRI-O implement the [`ConvertImage`][1] function to follow those self-defined

Reviewer 2: ```suggestion
CRI-O implements the [`ConvertImage`][1] function to follow the following self-defined
```

rules:

- always return at least one `repo_digests` value
- return zero or more `repo_tags` values

> There are multiple use cases where this behavior is relevant. Those will be

Reviewer 2: ```suggestion
There are multiple use-cases where this behavior is relevant. Those will be
```
